webrender/src/util.rs

```diff
@@ -8,12 +8,50 @@ use euclid::{Point2D, Rect, Size2D, TypedPoint2D, TypedRect, TypedSize2D, Vector
 use euclid::{TypedTransform2D, TypedTransform3D, TypedVector2D, TypedScale};
 use num_traits::Zero;
 use plane_split::{Clipper, Polygon};
-use std::{i32, f32, fmt};
+use std::{i32, f32, fmt, ptr};
 use std::borrow::Cow;
 
+
 // Matches the definition of SK_ScalarNearlyZero in Skia.
 const NEARLY_ZERO: f32 = 1.0 / 4096.0;
 
+/// A typesafe helper that separates new value construction from
+/// vector growing, which allows LLVM to elide the value copy.
+#[must_use]
+pub struct Allocation<'a, T: 'a> {
+    vec: &'a mut Vec<T>,
+    index: usize,
+}
+
+impl<'a, T> Allocation<'a, T> {
+    #[inline(always)]
+    pub fn init(self, value: T) -> usize {
```

Comment:
Maybe add:
// writing is safe because alloc() ensured enough capacity and Allocation holds a mutable borrow to prevent anyone else from breaking this invariant.

Reply:
Are you sure this actually avoids the memcpy / memmove? At least for `self` arguments a bit back we couldn't work around it like this, see https://github.com/rust-lang/rust/issues/42763.

Reply:
Confirmed in [playground](https://play.rust-lang.org/?version=stable&mode=release&edition=2015&gist=9aef0043bcb9a15155ae97119681f00d). In `fn foo()` changing the `push(xxx)` to `alloc().init(xxx)` removes the `memcpy`.

Reply:
I filed https://github.com/rust-lang/rust/issues/56333 for a problem contributing to the first example not working well in the playground.

Reply:
It looks like SmallVec also causes this to not work. I filed https://github.com/rust-lang/rust/issues/56356 about that.